tests/user_examples/extra_src/shared/general_continuum/continuum_dynamics/continuum_dynamics_complex.hpp

```diff
@@ -1,210 +1,103 @@
-//==========================BaseShearStressRelaxation1stHalfWithWall================================//
-//=================================================================================================//
-template <class BaseShearStressRelaxation1stHalfType>
-void BaseShearStressRelaxation1stHalfWithWall<BaseShearStressRelaxation1stHalfType>::interaction(size_t index_i, Real dt)
-{
-    BaseShearStressRelaxation1stHalfType::interaction(index_i, dt);
-    Matd shear_stress_i = this->shear_stress_[index_i];
-    Real rho_i = this->rho_[index_i];
-    Real rho_in_wall = this->continuum_.getDensity();
-    Vecd acceleration = Vecd::Zero();
-    for (size_t k = 0; k < FSIContactData::contact_configuration_.size(); ++k) // There may be several wall bodies.
+    //=================================================================================================//
+    namespace continuum_dynamics
     {
-        StdLargeVec<Vecd> &vel_ave_k = *(this->wall_vel_ave_[k]);
-        StdLargeVec<Vecd> &acc_ave_k = *(this->wall_acc_ave_[k]);
-        StdLargeVec<Vecd> &n_k = *(this->wall_n_[k]);
-        Neighborhood &wall_neighborhood = (*FSIContactData::contact_configuration_[k])[index_i];
-        for (size_t n = 0; n != wall_neighborhood.current_size_; ++n)
+        //=================================================================================================//
+        template <class BaseIntegrationType>
+        template <class BaseBodyRelationType, typename... Args>
+        InteractionWithWall<BaseIntegrationType>::
+            InteractionWithWall(BaseBodyRelationType& base_body_relation,
+                BaseContactRelation& wall_contact_relation, Args &&...args)
+            : BaseIntegrationType(base_body_relation, std::forward<Args>(args)...),
+            FSIContactData(wall_contact_relation)
         {
-            size_t index_j = wall_neighborhood.j_[n];
-            Vecd &e_ij = wall_neighborhood.e_ij_[n];
-            Real r_ij = wall_neighborhood.r_ij_[n];
-            Real dW_ijV_j = wall_neighborhood.dW_ijV_j_[n];
-            Vecd nablaW_ijV_j = wall_neighborhood.dW_ijV_j_[n] * wall_neighborhood.e_ij_[n];
-
-            Matd shear_stress_in_wall = shear_stress_i;
-            acceleration += this->rho_[index_j] * (shear_stress_i / (rho_i * rho_i) + shear_stress_in_wall / (rho_in_wall * rho_in_wall)) * nablaW_ijV_j;
+            if (&base_body_relation.getSPHBody() != &wall_contact_relation.getSPHBody())
+            {
+                std::cout << "\n Error: the two body_relations do not have the same source body!" << std::endl;
+                std::cout << __FILE__ << ':' << __LINE__ << std::endl;
+                exit(1);
+            }
+
+            for (size_t k = 0; k != FSIContactData::contact_particles_.size(); ++k)
+            {
+                Real rho0_k = FSIContactData::contact_bodies_[k]->base_material_->ReferenceDensity();
```

Comment:
Here, rho0_k will be re-assigned value during the iteration loop. More generally, a vector should be introduced?

Reply:
Seems ok, as rho0_k is temp variable.